### PR TITLE
Generate and publish code coverage reports in the CI

### PR DESCRIPTION
**Pull request by Dentosal**
Fuel-core counterpart for https://github.com/FuelLabs/fuel-vm/pull/750

This PR sets up code coverage reporting. It commits the report to [a branch called `codecov-pages`](https://github.com/FuelLabs/fuel-core/tree/codecov-pages), which contains a folder for each branch. Branch names are urlsafe-base64-encoded to avoid (security) bugs. Every time a PR workflow runs or the  `master` receives new commits, a CI action pushes updated codecov report there.

## Repository setup

The configuration has to be done by someone who has permission to edit [the repo settings](https://github.com/FuelLabs/fuel-core/settings/pages): set deployment from `codecov-pages` branch, like this:
Screenshot of the settings page

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -15,9 +15,11 @@ concurrency:
   cancel-in-progress: true
 
 env:
+  GIT_BRANCH: ${{ github.head_ref || github.ref_name }}
   CARGO_TERM_COLOR: always
   RUST_VERSION: 1.75.0
-  NIGHTLY_RUST_VERSION: nightly-2023-10-29
+  RUST_VERSION_FMT: nightly-2023-10-29
+  RUST_VERSION_COV: nightly-2024-06-05
```
Comment (MitchTurner) [inline .github/workflows/ci.yml:22]: Why do we use a different version of rust nightly?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -193,6 +195,48 @@ jobs:
           check-repo: false
           ignore-unpublished-changes: true
 
+  publish-codecov:
+    name: Publish code coverage report on GitHub pages branch
+    runs-on: buildjet-4vcpu-ubuntu-2204
+    needs:
+      - cargo-verifications
```
Comment (MitchTurner) [inline .github/workflows/ci.yml:202]: In the vm pr we have `needs: -cargo`. What's the difference?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Dentosal) [reply .github/workflows/ci.yml:22]: I didn't want to touch formatting in this PR. The branch coverage needs.a more recent nightly.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Dentosal) [reply .github/workflows/ci.yml:202]: This workflow doesn't have `cargo` job. The closes equivalent is `cargo-verifications`.

Comment (MitchTurner) [reply .github/workflows/ci.yml:202]: Okay. That makes sense. Do we have any issues for unifying them?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (MitchTurner) [reply .github/workflows/ci.yml:22]: Okay. That makes sense. Do we have any issues for unifying them?

**Comment by MitchTurner** [comment]
A lot of the links in the description were left as the `fuel-vm` links.

Also, when I tried to inspect the generated docs from the code cov branch I got this 404:
Screenshot 2024-06-14 at 11 42 22 AM

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Dentosal) [reply .github/workflows/ci.yml:22]: In a separate PR, maybe. I don't want to reformat the whole codebase here.

**Comment by Dentosal** [comment]
> A lot of the links in the description were left as the `fuel-vm` links.
> 
> Also, when I tried to inspect the generated docs from the code cov branch I got this 404:

Fixed description links. The github pages return 404 because the settings for this repo haven't been updated yet.

**Review by MitchTurner — APPROVED** [review]

**Pull request merged**